### DISCUSSION TITLE
How to pass additional paths defined in the workspace?

### DISCUSSION THREAD

**Discussion by bariod** [microsoft/vscode-isort#376] [Q&A] [answered]
I have defined multiple paths in the workspace (`python.analysis.extraPaths` and `terminal.integrated.env.windows`) to help with imports.
It seems isort isn't getting any of them: `[info] sys.path used to run Server` (in the extension output) is only listing the path(s) of the workspace interpreter - no additional ones.

 Is there a way to pass them to isort or is that not intended?

**Comment by karthiknadig** [comment]
@bariod Those are not intended for `isort`. Is there a particular scenario that is not working for you?

**Reply by bariod** [reply]
I have the following file structure (simplified):

```
root_folder
|-- flask
    |-- endpoints
        |-- module_a
            |-- endpoints.py
        |-- module_b
            |-- endpoints.py
        ..

    |-- util
        |-- util_file.py

|-- src
    |-- base
        |-- a
            |-- a.py

        |-- b
            |-- b.py

    |-- custom
```

Now, in `module_a.endpoints.py` I have the following imports (ordered correctly):

```
from datetime import datetime

from flask import Blueprint, make_response, render_template, request
from flask_babel import Domain

from a.a import ClassA
from util.util_file import UtilClass
```

And isort turns that into this:

```
from datetime import datetime

from a.a import ClassA
from flask import Blueprint, make_response, render_template, request
from flask_babel import Domain
from util.util_file import UtilClass
```

I'm running isort with the argument `--thirdparty=flask`, because of the local package's identical name.
I never want to import from the local `flask` package, so this should be fine.

To me, it seems that the local packages get treated as 3rd-party (whereas only `flask` and `flask_babel` actually are).

`python.analysis.extraPaths` and `terminal.integrated.env.windows` have the following paths:

- `"${workspaceFolder}/src/custom"`
- `"${workspaceFolder}/src/base"`
- `"${workspaceFolder}/src"`
- `"${workspaceFolder}/flask"`
- `"and/some/external/lib"`

Is my code relying to heavily on additional `PYTHONPATH`s?

**Reply by brettcannon** [reply]
> Is my code relying to heavily on additional `PYTHONPATH`s?

We prefer not to judge people and their choices, but your layout is not typical. Is there a reason you have so many distinct directories to hold your code? And any reason you are not using a virtual environment for external/3rd-party packages?

**Reply by bariod** [reply]
I'm not quite sure I get your question.  
There's only one external package that isn't installed via pip (thus the additional `"and/some/external/lib"`) the other paths are local folders. 

> you have so many distinct directories 

There's basically the `flask` one which holds REST-endpoints and then there's `src` containing the logic which is being called by the endpoints.

**Reply by brettcannon** [reply]
> the other paths are local folders

Right, but you're putting them all on to `sys.path` separately when most people would put those directories all in a single directory and have that be what had to be on `sys.path`. Otherwise you could do an editable install of all of your code into a virtual environment.

**Reply by bariod** [reply]
So do you reckon getting rid of the additional paths and adjusting all imports to only import from root level, isort would be able to "understand" the structure correctly?

**Reply by brettcannon** [reply, marked as answer]
It's more likely, but these days we recommend people use Ruff instead of isort.

**Reply by bariod** [reply]
Alright, thank you guys for your time :)
